### ISSUE TITLE
"Partitions must be sorted ascending..." error for sorted index

### ISSUE THREAD

**Issue by bnaul** [dask/dask#4461]
Versions:
```
numpy 1.16.0
pandas 0.24.0
dask 1.1.1
```

For a sorted index column w/ very far-apart values, I get an incorrect error about unsorted partitions. I see the same thing regardless of `shuffle` and `scheduler`; however, reducing the number of partitions or size of inputs can cause the error not to appear.

Example:
```
import numpy as np, pandas as pd, dask, dask.dataframe as dd

np.random.seed(0)
ids = sorted(np.random.randint(-960000000000000000, -950000000000000000, size=5000))
df1 = pd.DataFrame({'id': np.concatenate([np.repeat(_id, np.random.randint(1, 50)) for _id in ids]),
                   'x': 1})

ids = sorted(np.random.randint(-100000, 100000, size=5000))
df2 = pd.DataFrame({'id': np.concatenate([np.repeat(_id, np.random.randint(1, 50)) for _id in ids]),
                    'x': 2})
ids = sorted(np.random.randint(950000000000000000, 960000000000000000, size=5000))
df3 = pd.DataFrame({'id': np.concatenate([np.repeat(_id, np.random.randint(1, 50)) for _id in ids]),
                    'x': 3})

df = pd.concat((df1, df2, df3))
print('ID Monotonic:', df['id'].is_monotonic)
print('Setting index...')
with dask.config.set(shuffle='tasks', scheduler='single-threaded'):
    ddf = dd.from_pandas(df, npartitions=1000).set_index('id', sorted=True)
```
Output:
```
ID Monotonic: True
Setting index...
Traceback (most recent call last):
  File "index.py", line 21, in <module>
    ddf = dd.from_pandas(df, npartitions=1000).set_index('id', sorted=True)
  File "/Users/brett/venvs/model/lib/python3.7/site-packages/dask/dataframe/core.py", line 2677, in set_index
    **kwargs)
  File "/Users/brett/venvs/model/lib/python3.7/site-packages/dask/dataframe/shuffle.py", line 511, in set_sorted_index
    divisions = compute_divisions(result, **kwargs)
  File "/Users/brett/venvs/model/lib/python3.7/site-packages/dask/dataframe/shuffle.py", line 535, in compute_divisions
    mins, maxes)
ValueError: ('Partitions must be sorted ascending with the index', 0     -959996520943834061
1     -959987606102464091
2     -959978667928021619
              ...
995   -950037686828763902
996   -950021743008315948
997                 98451
998                 99161
999                 99651
Length: 1000, dtype: int64, 0      950005961189076413
1      950018444802560903
2      950025599822840051
              ...
994    959999375588005446
995    959999375588005446
996                 98451
997                 99161
998                 99651
999                 99934
Length: 1000, dtype: int64)
```

**Comment by mrocklin** [comment]
Thanks @bnaul .  I agree that that looks like an issue.  Any chance it's just integer overflow ?  

As a disclaimer, I personally don't have time to investigate this, so I wouldn't encourage you to wait on me.  Ideally someone else can help out, maybe you?

**Comment by bnaul** [comment]
Same thing w/ smaller ints, but I do think I got to the bottom of it. Here's a peek into the traceback:
```
> /Users/brett/venvs/model/lib/python3.7/site-packages/dask/dataframe/core.py(2677)set_index()
   2675             from .shuffle import set_sorted_index
   2676             return set_sorted_index(self, other, drop=drop, divisions=divisions,
-> 2677                                     **kwargs)
   2678         else:
   2679             from .shuffle import set_index

ipdb> self
Dask DataFrame Structure:
                    id      x
npartitions=999
0                int64  int64
127                ...    ...
...                ...    ...
126586             ...    ...
126661             ...    ...
Dask Name: from_pandas, 999 tasks
ipdb> self.head()
        id  x
0 -9599990  1
0  9500027  3
0    -1000  2
1 -9599990  1
1  9500027  3
```
Looks like the issue is the `sort=True` default argument to `dd.from_pandas`, which I did not know about 🙃 .

**Comment by bnaul** [comment]
Closing for now but the original place I encountered this behavior did not use `dd.from_pandas`, so stay tuned...

**Comment by bnaul** [comment]
I dug into this a little more; the original issue had to do with partitions not respecting divisions, and the above was an attempt at an example but I think missed the actual issue.

New example:
```
import numpy as np, pandas as pd, dask, dask.dataframe as dd

np.random.seed(0)
ids = sorted(np.random.randint(-9600000, -9500000, size=5000))
df1 = pd.DataFrame({'id': np.concatenate([np.repeat(_id, np.random.randint(1, 50)) for _id in ids]),
                   'x': 1})

ids = sorted(np.random.randint(-1000, 1000, size=5000))
df2 = pd.DataFrame({'id': np.concatenate([np.repeat(_id, np.random.randint(1, 50)) for _id in ids]),
                    'x': 2})
ids = sorted(np.random.randint(9500000, 9600000, size=5000))
df3 = pd.DataFrame({'id': np.concatenate([np.repeat(_id, np.random.randint(1, 50)) for _id in ids]),
                    'x': 3})

df = pd.concat((df1, df2, df3))
with dask.config.set(shuffle='disk', scheduler='single-threaded'):
    ddf = dd.from_pandas(df, npartitions=173, sort=False).set_index('id', sorted=True)
    print(ddf.get_partition(1).tail())
    print(ddf.get_partition(2).head())
```
Output:
```
          x
id
-9596591  1
-9596591  1
-9596582  1
-9596582  1
-9596582  1
          x
id
-9596582  1
-9596582  1
-9596582  1
-9596582  1
-9596582  1
```
My understanding was that, once `set_index` is called, it's safe to assume that all values w/ a certain index are in the same partition. That does seem to be the case if we explicitly call `.repartition` w/ the same divisions:
```
In [5]: repart = ddf.repartition(divisions=ddf.divisions)

In [6]: repart.get_partition(1).tail()
Out[6]:
          x
id
-9599391  1
-9599391  1
-9599391  1
-9599391  1
-9599391  1

In [7]: repart.get_partition(2).head()
Out[7]:
          x
id
-9599388  1
-9599388  1
-9599388  1
-9599388  1
-9599388  1
```
@mrocklin would you say this counts as a bug or just undefined behavior for duplicate index values...? If it's the former I'll open a new issue.

**Comment by mrocklin** [comment]
>     ddf = dd.from_pandas(df, npartitions=173, sort=False).set_index('id', sorted=True)

Note that the `sorted=` keyword isn't perscriptive.  You're not saying "and please sort these".  Instead you're making a guarantee that they are sorted.  I would guess that if you remove that keyword that things would be closer to what you want.

**Comment by mrocklin** [comment]
To be clear, you're making a guarantee that they are *already* sorted, before the `set_index` operation, which in this case may not be true.

**Comment by bnaul** [comment]
Sorry, I think my example was unclear / complicated by the multiple attempts: by construction the dataframe index in the example is already sorted. There was a print statement asserting that in the first example ("monotonic: True") but it got lost in the repaste. So the behavior I'm seeing is:
- new index column is pre-sorted
- set_index(sorted=True) sets the index w/o complaining
- ddf's index is still ordered, but repeated values are sometimes split between adjacent partitions

I don't see any treatment of the duplicate index case in the docs so this might be just an ambiguity, but since that's the recommended pattern for grouping+applying this seems undesirable. `repartition` does have the behavior I would expect, combining all of the rows for a key in the same partition.

**Comment by mrocklin** [comment]
Right, ok yes that makes sense.  I suppose that the `set_index(..., sorted=True)` option is really assuming that the index is sorted and that values are cleanly partitioned.  You're right though that that isn't at all obvious, nor is it easy for a user to verify/ensure that condition (unless they know that their values are likely unique).

**Comment by cgi1** [comment]
> To be clear, you're making a guarantee that they are _already_ sorted, before the `set_index` operation, which in this case may not be true.

Understood.. Well, thats a problem if you have 99,999% correctly sorted data. Than you need to reshuffle the whole set.

**Comment by TomAugspurger** [comment]
What are you suggesting @cgi1?

> Than you need to reshuffle the whole set.

If your data are only unsorted within partitions, then the you don't need a shuffle. You would only need to sort each partition independently.

**Comment by cgi1** [comment]
@TomAugspurger So what I came up with is [this little cleanup script for large csvs](https://github.com/cgi1/Cheats/blob/master/sort_large_csv_by_timestamp_using_pandas.py). Afterwards the `sorted=True` import is actually working.  Anyway, the question arises, if this kind of error behaviour could somehow be integrated into the `set_index(.., sorted=True)` function of `dask`.  In my example case, I found out, that I had errors in only 2 of 76.661.682 rows of data and some rows which are very closely related timely (e.g. a few microseconds before the previous row). 

> If your data are only unsorted within partitions, then the you don't need a shuffle. You would only need to sort each partition independently.

So basically, the data is sorted correctly in 99.9999% of the cases.  It would be an absolut edge case if the 0.00001% would match exactly on the boundaries of the partitions. 

Anyway, I could `df.set_index()` for each partition and than run a `set_index(sorted=True)` on the whole `ddf`? Is that what you are saying? And that would be faster than a complete `ddf.set_index()`?

**Comment by TomAugspurger** [comment]
>  Anyway, the question arises, if this kind of error behaviour could somehow be integrated into the set_index(.., sorted=True)

I'm not sure. Currently `sorted=True` has the meaning that this data is pre-sorted, so Dask doesn't need to load data to compute the percentiles for discovering partitions. I'm hesitant to complicate that meaning at all.

If you're comfortable with the (small) risk of out-of-order data being selected for a partition bound, I'd recommend

```python
df = df.set_index("column", sorted=True).map_partitions(lambda x: x.sort_index())`
```
